### PR TITLE
refactor!: Seal the Dependency class

### PR DESCRIPTION
**Pull request by spydon**
- Thanks for your contribution! Please replace this text with a description of what this PR is changing or adding and why, list any relevant issues, and review the contribution guidelines below.

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

Seals the `Dependency` class to make it easier to pattern match over the types.

### CODE REVIEW

**Comment by kevmoo** [comment]
changelog, please!

**Comment by kevmoo** [comment]
Also, we'll have to make this a breaking change to be paranoid

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ## 1.3.1-wip
```
Comment (kevmoo) [inline CHANGELOG.md:1]: Pretty sure we'll have to make this v2.0.0 – since it's a breaking change

**Comment by spydon** [comment]
> Also, we'll have to make this a breaking change to be paranoid

Since it had a private constructor before this change nobody could extend it, so I don't think we'd have to mark this as breaking?

File: CHANGELOG.md (hunk 1 above)
Comment (spydon) [reply CHANGELOG.md]: Done!

**Comment by kevmoo** [comment]
> > Also, we'll have to make this a breaking change to be paranoid
> 
> Since it had a private constructor before this change nobody could extend it, so I don't think we'd have to mark this as breaking?
> 
> EDIT: Ah, but they could have extended the subclasses...

This is the annoying bit. It's like 99% likely to NOT be a problem...but...

**Comment by kevmoo** [comment]
@devoncarew @natebosch @jakemac53 – thoughts on breaking change or not here?

**Comment by spydon** [comment]
@kevmoo @natebosch @jakemac53 @devoncarew can we land this PR and #137 before moving the repo?

**Comment by kevmoo** [comment]
I'm waiting for a second opinion on sealing the classes being breaking...

**Comment by spydon** [comment]
> I'm waiting for a second opinion on sealing the classes being breaking...

Isn't there some google stick you can poke them with? :sweat_smile: 

Right, now we get this of course:

> Because pubspec_parse depends on json_serializable ^6.6.0 which depends on pubspec_parse ^1.0.0, pubspec_parse ^1.0.0 is required.

Imho I don't think we'd need to do a breaking change for this change, it's do incredibly unlikely that it would cause any breakage for anyone. I'll set it back to 1.3.1, and if you disagree I'll make it 2.0.0 again.

**Comment by devoncarew** [comment]
> Imho I don't think we'd need to do a breaking change for this change, it's do incredibly unlikely that it would cause any breakage for anyone. I'll set it back to 1.3.1, and if you disagree I'll make it 2.0.0 again.

I guess the argument is that people couldn't have subclassed Dependency, they could have subclasses SdkDependency, PathDependency, ..., but that it's unlikely that people would have actually done that? So technically a breaking change but one that's unlikely to break anyone in practice.

Do we know if the main users (a few of https://pub.dev/packages?q=dependency%3Apubspec_parse) are safe w/ this change?

**Comment by spydon** [comment]
@devoncarew I went through the two first pages (based on likes) and none of them extends any of the `Dependency` subclasses.

**Comment by natebosch** [comment]
Yeah I think it would be OK to land this without a major version change. We should switch to a feature version bump though.

File: CHANGELOG.md (hunk 1 above)
Comment (natebosch) [inline CHANGELOG.md:1]:
```suggestion
## 1.4.0-wip
```

**Review by natebosch — COMMENTED** [review]
We will need to update the pubpsec too.

**Comment by spydon** [comment]
@natebosch bumped to 1.4.0

**Review by kevmoo — APPROVED** [review]

**Comment by kevmoo** [comment]
Thanks @spydon !

**Pull request merged**